be/src/formats/csv/csv_reader.cpp

```diff
@@ -2,20 +2,566 @@
+            _buff.skip(_column_separator_length);
+            return true;
+        }
+    }
+    return false;
+}
+
+inline bool CSVReader::isRowDelimiter(bool expandBuffer) {
+    if (LIKELY(_row_delimiter_length == 1)) {
+        if (*(_buff.position()) == _parse_options.row_delimiter[0]) {
+            _buff.skip(1);
+            return true;
+        }
+    } else {
+        int i = 0;
+        const char* base_ptr = _buff.base_ptr();
+        size_t p = _buff.position_offset();
+        while (i < _row_delimiter_length && p < _buff.limit_offset() &&
+               *(base_ptr + p) == _parse_options.row_delimiter[i]) {
+            i++;
+            p++;
+            if (_buff.limit_offset() - p < 1) {
+                Status s = readMore(expandBuffer);
+                if (!s.ok()) {
+                    return false;
+                }
+                if (_buff.limit_offset() - p < 1) {
+                    return false;
+                }
+            }
+        }
+        if (i == _row_delimiter_length) {
+            _buff.skip(_row_delimiter_length);
+            return true;
+        }
+    }
+    return false;
+}
+
+char* CSVReader::buffBasePtr() {
+    return _buff.base_ptr();
+}
+
+char* CSVReader::escapeDataPtr() {
+    return _escape_data.data();
+}
+
+Status CSVReader::buffInit() {
+#ifndef CSVBENCHMARK
+    _buff.compact();
+    if (_buff.free_space() > 0) {
+        return _fill_buffer();
+    }
+#endif
+    return Status::OK();
+}
+
+Status CSVReader::readMore(bool expandBuffer) {
+    if (_buff.free_space() > 0) {
+        return _fill_buffer();
+    } else {
+        if (expandBuffer) {
+            Status s = _expand_buffer_loosely();
+            if (!s.ok()) {
+                return s;
+            }
+            return _fill_buffer();
+        } else {
+            return Status::OK();
+        }
+    }
+}
+
+Status CSVReader::more_lines() {
+    if (UNLIKELY(_limit > 0 && _parsed_bytes > _limit)) {
+        return Status::EndOfFile("Reached limit");
+    }
+    Status status = Status::OK();
+    status = buffInit();
+    if (!status.ok()) {
+        return status;
+    }
+    _escape_data.clear();
+    _escape_pos.clear();
+    ParseState curState = START;
+    ParseState preState = curState;
+    // parsed_start and parsed_end record a line at the start and end of the buff.
+    size_t parsed_start = _buff.position_offset();
+    size_t parsed_end = 0;
+    // If filed_start is initialized to the maximum value, the START state is not entered.
+    size_t filed_start = std::string::npos;
+    size_t filed_end = 0;
+    bool is_enclose_field = false;
+    bool is_escape_field = false;
+    bool notGetLine = true;
+    bool reachBuffEnd = false;
+    _fields.clear();
+    while (true) {
+        // At the end of a row, or the end of a field, no new data is read.
+        if (LIKELY(curState != NEWLINE && curState != DELIMITER)) {
+            if (UNLIKELY(_buff.available() < 1)) {
```

Comment:
How about adding a function to read data on demand to simplify this?

Reply:
Yes, we can use a macro to simplify this.